MapleServer2/PacketHandlers/Game/MeretMarketHandler.cs

```diff
@@ -55,27 +90,274 @@ public override void Handle(GameSession session, PacketReader packet)
-        List<MeretMarketItem> marketItems = DatabaseManager.MeretMarket.FindAllByCategoryId(category);
-        if (marketItems == null)
+
+        MeretMarketCategoryMetadata metadata = MeretMarketCategoryMetadataStorage.GetMetadata((int) category);
+        if (metadata is null)
+        {
+            return;
+        }
+
+        switch (metadata.Section)
+        {
+            case MeretMarketSection.PremiumMarket:
+                HandleOpenPremiumMarket(session, category);
+                break;
+            case MeretMarketSection.RedMeretMarket:
+                HandleOpenRedMeretMarket();
+                break;
+            case MeretMarketSection.UGCMarket:
+                HandleOpenUGCMarket(session, packet, metadata);
+                break;
+        }
+    }
+
+    private static void HandleOpenPremiumMarket(GameSession session, MeretMarketCategory category)
+    {
+        List<MeretMarketItem> marketItems = DatabaseManager.MeretMarket.FindAllByCategoryId((category));
+        if (marketItems is null)
         {
             return;
         }
-        session.Send(MeretMarketPacket.Premium(marketItems));
+        session.Send(MeretMarketPacket.LoadPremiumShopCategory(marketItems));
+    }
+
+    private static void HandleOpenUGCMarket(GameSession session, PacketReader packet, MeretMarketCategoryMetadata metadata)
+    {
+        GenderFlag gender = (GenderFlag) packet.ReadByte();
+        JobFlag job = (JobFlag) packet.ReadInt();
+        short sortBy = packet.ReadByte();
+
+        List<UGCMarketItem> items = GameServer.UGCMarketManager.FindItemsByCategory(metadata.ItemCategories, gender, job, sortBy);
+        session.Send(MeretMarketPacket.LoadUGCShopCategory(items));
+    }
+
+    private static void HandleOpenRedMeretMarket()
+    {
+        // TODO: Red Meret Market
     }
 
     private static void HandlePurchase(GameSession session, PacketReader packet)
     {
         byte quantity = packet.ReadByte();
         int marketItemId = packet.ReadInt();
-        byte[] unk = packet.ReadBytes(12);
+        long ugcItemId = packet.ReadLong();
+        if (ugcItemId != 0)
+        {
+            PurchaseUGCItem(session, ugcItemId);
+            return;
+        }
+
+        PurchasePremiumItem(session, packet, marketItemId);
+        return;
+    }
+
+    private static void PurchaseUGCItem(GameSession session, long ugcMarketItemId)
+    {
+        UGCMarketItem marketItem = GameServer.UGCMarketManager.FindItemById(ugcMarketItemId);
+        if (marketItem is null || marketItem.ListingExpirationTimestamp < TimeInfo.Now())
+        {
+            return;
+        }
+
+        if (!HandleMarketItemPay(session, marketItem.Price, MeretMarketCurrencyType.Meret))
+        {
+            return;
+        }
+
+        marketItem.SalesCount++;
+        DatabaseManager.UGCMarketItems.Update(marketItem);
+        UGCMarketSale sale = new(marketItem.Price, marketItem.Item.UGC.Name, marketItem.SellerCharacterId);
```

Comment:
not being used

Reply:
What do you mean? it makes a new `UGCMarketSale` and within the constructer it adds it to the `UGCMarketManager`

Reply:
I meant the variable is not being used